Add functionality to custom study by tags or card state

lukstbit opened the pull request:
## Purpose / Description
Enables for the app the full desktop functionality related to custom study by tags. Some notes related to the implementation:

- follows the  desktop flow with "Choose tags" button showing a new dialog and on ok on that dialog returning to the previous dialog to do the actual setup of the custom study
- uses different layouts for portrait/landscape but the transition between them doesn't work because DeckPicker is set(although it shouldn't) to handle the configuration manually
- I added a border for both lists because the middle label was kind of blending with the lists IMO


<details>
<summary>How it looks on mobile</summary>


![Screenshot from 2025-02-09 20-11-15](https://github.com/user-attachments/assets/62621769-d918-4501-9d67-e09884c63d29)
![Screenshot from 2025-02-09 20-11-45](https://github.com/user-attachments/assets/0e489a4d-06d5-415a-a3df-c5c33476beb0)![Screenshot from 2025-02-09 20-14-58](https://github.com/user-attachments/assets/4b7b3feb-cf54-46c3-a65d-ae4570e86d7d)


![Screenshot from 2025-02-09 20-15-52](https://github.com/user-attachments/assets/ceeee637-25ea-4945-98e2-d958b41f6157)

![Screenshot from 2025-02-09 20-12-57](https://github.com/user-attachments/assets/71d27917-9d73-4a24-b5f0-f3eda60ec57e)


![Screenshot from 2025-02-09 20-17-13](https://github.com/user-attachments/assets/9931da51-ed4c-418b-9e9e-25a8d37bdc03)
![Screenshot from 2025-02-09 20-16-33](https://github.com/user-attachments/assets/d5414efd-41be-4525-8546-5a6c956c3c1d)

</details>

<details>
<summary>How it looks on desktop</summary>

![Screenshot from 2025-02-09 20-51-31](https://github.com/user-attachments/assets/a488f7c2-e4da-4bd7-bb73-4dd557c84bb0)
![Screenshot from 2025-02-09 20-51-41](https://github.com/user-attachments/assets/64d8ebda-ef7d-47cf-b6df-cf1082dbe80b)

</details>


## Fixes
* Fixes #17583
* Fixes #17984

## How Has This Been Tested?
Ran the tests, manually verified the new dialogs/code paths.

## Checklist

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt:
```diff
@@ -421,70 +424,12 @@ class CustomStudyDialog :
                 STUDY_AHEAD -> prefs.getInt("aheadDays", 1).toString()
                 STUDY_PREVIEW -> prefs.getInt("previewDays", 1).toString()
-                STUDY_TAGS,
+                STUDY_TAGS -> prefs.getInt("amountOfCards", 100).toString()
```
Could you document that this constant isn't upstream as-of Anki 25.02

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt:
```diff
@@ -312,6 +336,9 @@ class CustomStudyDialog :
     private suspend fun customStudy(
         contextMenuOption: ContextMenuOption,
         userEntry: Int,
+        cramKind: CramKind = CramKind.CRAM_KIND_NEW,
```
nit: this feels like it should exist on `ContextMenuOption.STUDY_TAGS`, as they're only relevant to one branch

I suspect it's not worth recoding this now, unless it's easy, could a refactoring TODO be added?

david-allison replied:
Very likely needs sealed classes, happy to leave this

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/IncludedExcludedTagsAdapter.kt:
```diff
@@ -0,0 +1,131 @@
+import com.ichi2.anki.utils.ext.findViewById
+
+/**
+ * Shows a simple list of tags from which the user can select for his custom study session. For
```
Sorry not sorry

```suggestion
 * Shows a simple list of tags from which the user can select for a custom study session. For
```

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/IncludedExcludedTagsAdapter.kt:
```diff
@@ -0,0 +1,131 @@
+ * simplicity this adapter is used for both types of lists.
+ *
+ * @param mode determines how the tags in the list backing this adapter are to be handled(either as
+ * included or excluded tags for the custom study)
```
It would be useful to link the 'return value' of the adapter with an `@see`

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/IncludedExcludedTagsAdapter.kt:
```diff
@@ -0,0 +1,131 @@
+    private val selectedItemBackground: Int
+
+    init {
+        var ta = context.obtainStyledAttributes(intArrayOf(android.R.attr.selectableItemBackground))
+        selectableItemBackground = ta.getResourceId(0, 0)
+        ta.recycle()
+        ta = context.obtainStyledAttributes(intArrayOf(com.google.android.material.R.attr.colorPrimary))
+        selectedItemBackground = ta.getColor(0, Color.BLUE)
+        ta.recycle()
```
... or possibly kept as slightly less efficient, and moved to the variables themselves

lukstbit replied:
Combined them together for now.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/TagLimitFragment.kt:
```diff
@@ -0,0 +1,163 @@
+
+    override fun onStart() {
+        super.onStart()
+        val deckId = requireArguments().getLong(ARG_DECK_ID)
```
would prefer this to be a property. Implementer's choice

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/TagLimitFragment.kt:
```diff
@@ -0,0 +1,163 @@
+ * Fragment that allows the user to select tags to include and/or exclude for the custom study
+ * session that is being creating. Similar with the desktop TagLimit class.
+ * https://github.com/ankitects/anki/blob/main/qt/aqt/taglimit.py#L17
+ * @see CustomStudyDialog
```
I'd also find linking the 'return value' here to be useful

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/TagLimitFragment.kt:
```diff
@@ -0,0 +1,163 @@
+
+/**
+ * Fragment that allows the user to select tags to include and/or exclude for the custom study
+ * session that is being creating. Similar with the desktop TagLimit class.
```
```suggestion
 * session that is being created. Similar to the desktop TagLimit class.
```

david-allison reviewed (changes requested):
This is great! Don't let any of these nitpicks get in the way of getting it in. It's more than 'good enough'

Mostly UI nitpicks:

Margins are off:

<img width="41" alt="Screenshot 2025-02-13 at 09 10 08" src="https://github.com/user-attachments/assets/686d1768-4fc1-4449-b08f-4cd3362a26ed" />

* add a little more spacing between the 'Select tags to exclude' and the list
* if there is a scrollbar, always show it in the selection dialog
* make the 'exclude' label look like a title
* Anki replaces `"_"` with `" "` when visually displaying tags in this screen. Optional if you want to do this
* (optional/add as TODO): it feels like we want a ripple select
* I /think/ the highlighting should go away if a user unchecks the checkbox

lukstbit commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt:
```diff
@@ -259,14 +271,20 @@ class CustomStudyDialog :
         val dialog =
             AlertDialog
                 .Builder(requireActivity())
                 .customView(view = v, paddingStart = 64, paddingEnd = 64, paddingTop = 32, paddingBottom = 32)
```
Fixed, see the images.

lukstbit commented:
Implemented most requests, except for:

_I /think/ the highlighting should go away if a user unchecks the checkbox_

because that's how desktop behaves.

<details>

<summary>How it looks now</summary>

![Screenshot from 2025-02-14 12-11-08](https://github.com/user-attachments/assets/9c826254-2761-425e-a72d-6783b12faa8b)
![Screenshot from 2025-02-14 12-11-32](https://github.com/user-attachments/assets/a0538024-8a15-417c-ba07-93362499484c)
![Screenshot from 2025-02-14 12-11-49](https://github.com/user-attachments/assets/05fbd987-99cb-4159-bc1c-1780c41f7318)
![Screenshot from 2025-02-14 12-21-28](https://github.com/user-attachments/assets/9848e299-1d36-4f36-b44a-e5e2bade4374)
![image](https://github.com/user-attachments/assets/eeb91604-fc6d-4916-9b31-c371aeaef04c)

</details

david-allison commented:
Select tags to exclude doesn't strike me as a title yet. Vertical pudding feels a little off, everything else looks great

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/IncludedExcludedTagsAdapter.kt:
```diff
@@ -0,0 +1,138 @@
+    private val selectableItemBackground: Int
+
+    /** Background color for a tag that is selected(references R.attr.colorPrimary). */
+    // TODO implement a ripple effect or combine the two backgrounds into one background drawable
```
Cheers!

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/IncludedExcludedTagsAdapter.kt:
```diff
@@ -0,0 +1,138 @@
+        position: Int,
+    ) {
+        val model = tags[position]
+        holder.tagView.text = model.name.replace("_", " ")
```
I'd be tempted to name this `label`, or `userFacingLabel` on the class

david-allison reviewed (approved):
I retract my past statement:

* I would like a minor re-think of the design of "select tags to exclude", as it still doesn't fully feel like a title

lukstbit commented:
Small changes in the latest push:
- rename the previous _fragment_custom_study_tags_selection_ layouts to _fragment_tag_limit_.
- fixed a bug in CustomStudyDialog's positive action(see comment there): If we went to TagLimit and we cancelled, in the previous CustomStudyDialog we wouldn't be able to start again TagLimit
- added a userFacingLabel property for the model TagIncludedExcluded
- used _?textAppearanceTitleMedium_ for _"Select tags to exclude"_ label:
![Screenshot from 2025-02-18 18-53-38](https://github.com/user-attachments/assets/3fbdffcd-8bf4-4159-98f3-e2a37a177c3a)

mikehardy reviewed (approved):
Wow, heck of a feature, thank you!
I have synced strings so this should be good to go
